File: python/src/vcard/parser.py
"""
vCard Parser for RFC 6350
"""

from typing import List
from .dom import VCardObject, VCardProperty


class ParseException(Exception):
    """Exception raised when parsing fails"""
    pass


class VCardParser:
    """Parser for text/vcard format (RFC 6350)"""

    def __init__(self):
        self._lines: List[str] = []
        self._current_line: int = 0

    def parse(self, vcard_text: str) -> List[VCardObject]:
        """Parse vCards from text (returns all vCards found)"""
        vcards = []
        self._lines = self._unfold_lines(vcard_text)
        self._current_line = 0

        while self._current_line < len(self._lines):
            # Skip empty lines
            while self._current_line < len(self._lines) and not self._lines[self._current_line].strip():
                self._current_line += 1

            if self._current_line >= len(self._lines):
                break

            line = self._lines[self._current_line]
            if line.upper() == "BEGIN:VCARD":
                self._current_line += 1
                vcard = VCardObject()
                self._parse_component(vcard)
                vcards.append(vcard)
            else:
                raise ParseException(f"Expected BEGIN:VCARD but got: {line}")

        if not vcards:
            raise ParseException("No vCard data found")

        return vcards

    def parse_file(self, file_path: str) -> List[VCardObject]:
        """Parse vCards from a file"""
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        return self.parse(content)

    def _unfold_lines(self, vcard_text: str) -> List[str]:
        """Unfold long lines according to RFC 6350"""
        unfolded_lines = []
        lines = vcard_text.replace('\r\n', '\n').split('\n')

        current_line = []

        for line in lines:
            if line and (line[0] == ' ' or line[0] == '\t'):
                # Continuation line - remove leading whitespace and append
                current_line.append(line[1:])
            else:
                if current_line:
                    unfolded_line = ''.join(current_line)
                    if unfolded_line.strip():
                        unfolded_lines.append(unfolded_line)
                current_line = [line]

        if current_line:
            unfolded_line = ''.join(current_line)
            if unfolded_line.strip():
                unfolded_lines.append(unfolded_line)

        return unfolded_lines

    def _parse_component(self, component: VCardObject):
        """Parse properties of a component"""
        while self._current_line < len(self._lines):
            line = self._lines[self._current_line]

            if line.upper().startswith("END:"):
                end_component_type = line[4:].upper()
                if end_component_type != component.component_type:
                    raise ParseException(
                        f"Mismatched END tag: expected END:{component.component_type} but got END:{end_component_type}"
                    )
                self._current_line += 1

                # Validate required properties for vCard
                self._validate_vcard(component)

                return
            else:
                prop = self._parse_property(line)
                component.add_property(prop)
                self._current_line += 1

        raise ParseException(f"Unexpected end of input while parsing {component.component_type}")

    def _parse_property(self, line: str) -> VCardProperty:
        """Parse a property line"""
        colon_index = self._find_unquoted_char(line, ':')
        if colon_index == -1:
            raise ParseException(f"Invalid property line (missing colon): {line}")

        name_and_params = line[:colon_index]
        value = line[colon_index + 1:]

        # Unescape value
        value = self._unescape_value(value)

        # Parse name and parameters
        semicolon_index = self._find_unquoted_char(name_and_params, ';')

        if semicolon_index != -1:
            property_name = name_and_params[:semicolon_index].upper()
            params_part = name_and_params[semicolon_index + 1:]
        else:
            property_name = name_and_params.upper()
            params_part = None

        prop = VCardProperty(property_name, value)

        if params_part:
            self._parse_parameters(params_part, prop)

        return prop

    def _parse_parameters(self, params_part: str, prop: VCardProperty):
        """Parse property parameters"""
        parameters = self._split_parameters(params_part)

        for param in parameters:
            equals_index = param.find('=')
            if equals_index == -1:
                raise ParseException(f"Invalid parameter (missing equals): {param}")

            param_name = param[:equals_index].upper()
            param_value = param[equals_index + 1:]

            # Remove quotes if present
            if param_value.startswith('"') and param_value.endswith('"') and len(param_value) >= 2:
                param_value = param_value[1:-1]

            # Handle comma-separated values
            values = self._split_parameter_values(param_value)
            for value in values:
                prop.add_parameter(param_name, value)

    def _split_parameters(self, params_part: str) -> List[str]:
        """Split parameters by semicolon, respecting quotes"""
        parameters = []
        current = []
        in_quotes = False

        for c in params_part:
            if c == '"':
                in_quotes = not in_quotes
                current.append(c)
            elif c == ';' and not in_quotes:
                parameters.append(''.join(current))
                current = []
            else:
                current.append(c)

        if current:
            parameters.append(''.join(current))

        return parameters

    def _split_parameter_values(self, param_value: str) -> List[str]:
        """Split parameter values by comma, respecting quotes"""
        values = []
        current = []
        in_quotes = False

        for c in param_value:
            if c == '"':
                in_quotes = not in_quotes
            elif c == ',' and not in_quotes:
                values.append(''.join(current))
                current = []
            else:
                current.append(c)

        if current:
            values.append(''.join(current))

        return values

    def _find_unquoted_char(self, s: str, target: str) -> int:
        """Find the first occurrence of target not inside quotes"""
        in_quotes = False
        for i, c in enumerate(s):
            if c == '"':
                in_quotes = not in_quotes
            elif c == target and not in_quotes:
                return i
        return -1

    def _unescape_value(self, value: str) -> str:
        """Unescape special characters in value"""
        return (
            value
            .replace('\\n', '\n')
            .replace('\\N', '\n')
            .replace('\\;', ';')
            .replace('\\,', ',')
            .replace('\\\\', '\\')
        )

    def _validate_vcard(self, vcard: VCardObject):
        """Validate required vCard properties"""
        # VERSION is required (RFC 6350 Section 6.7.9)
        if not vcard.version:
            raise ParseException(
                "Missing required VERSION property (RFC 6350 Section 6.7.9). "
                "vCard must include VERSION:4.0"
            )

        # Only version 4.0 is supported (per ADR 0004)
        if vcard.version != "4.0":
            raise ParseException(
                f"Unsupported vCard version: {vcard.version}. "
                "Only version 4.0 is supported (see ADR 0004)."
            )

        # FN (formatted name) is required (RFC 6350 Section 6.2.1)
        if not vcard.formatted_name:
            raise ParseException(
                "Missing required FN (Formatted Name) property (RFC 6350 Section 6.2.1). "
                "vCard must include FN property."
            )

File: typescript/src/parser.ts
/**
 * Parser for text/vcard format (RFC 6350)
 */

import { VCardComponent, VCardObject, VCardProperty } from './dom';
import { Telephone, Email, Address } from './types';

export class ParseException extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ParseException';
  }
}

export class VCardParser {
  private lines: string[] = [];
  private currentLine: number = 0;

  /**
   * Parse vCard text content. Always returns a list (per ADR 0001).
   */
  parse(vcardText: string): VCardObject[] {
    const vcards: VCardObject[] = [];
    this.lines = this.unfoldLines(vcardText);
    this.currentLine = 0;

    while (this.currentLine < this.lines.length) {
      // Skip empty lines
      while (this.currentLine < this.lines.length && this.lines[this.currentLine].trim() === '') {
        this.currentLine++;
      }
      if (this.currentLine >= this.lines.length) break;

      const line = this.lines[this.currentLine];
      if (line.toUpperCase() === 'BEGIN:VCARD') {
        this.currentLine++;
        const vcard = new VCardObject();
        this.parseComponent(vcard);
        vcards.push(vcard);
      } else {
        throw new ParseException(`Expected BEGIN:VCARD but got: ${line}`);
      }
    }

    if (vcards.length === 0) {
      throw new ParseException('No vCard data found');
    }

    return vcards;
  }

  /**
   * Parse vCard from file content (convenience wrapper for Node.js usage)
   */
  parseFile(filePath: string): VCardObject[] {
    const fs = require('fs');
    const content = fs.readFileSync(filePath, 'utf-8');
    return this.parse(content);
  }

  private unfoldLines(vcardText: string): string[] {
    const unfoldedLines: string[] = [];
    const rawLines = vcardText.split(/\r\n|\n/);
    let currentLine = '';

    for (const line of rawLines) {
      if (line.length > 0 && (line[0] === ' ' || line[0] === '\t')) {
        // Continuation line
        currentLine += line.substring(1);
      } else {
        if (currentLine.trim() !== '') {
          unfoldedLines.push(currentLine);
        }
        currentLine = line;
      }
    }

    if (currentLine.trim() !== '') {
      unfoldedLines.push(currentLine);
    }

    return unfoldedLines;
  }

  private parseComponent(component: VCardComponent): void {
    while (this.currentLine < this.lines.length) {
      const line = this.lines[this.currentLine];

      if (line.toUpperCase().startsWith('END:')) {
        const endComponentType = line.substring(4).toUpperCase();
        if (endComponentType !== component.componentType) {
          throw new ParseException(
            `Mismatched END tag: expected END:${component.componentType} but got END:${endComponentType}`,
          );
        }
        this.currentLine++;

        if (component instanceof VCardObject) {
          this.validateVCard(component);
        }
        return;
      } else {
        const property = this.parseProperty(line);

        if (component instanceof VCardObject) {
          switch (property.name) {
            case 'TEL':
              component.telephones.push(Telephone.fromProperty(property));
              component.addProperty(property);
              break;
            case 'EMAIL':
              component.emails.push(Email.fromProperty(property));
              component.addProperty(property);
              break;
            case 'ADR':
              component.addresses.push(Address.fromProperty(property));
              component.addProperty(property);
              break;
            default:
              component.addProperty(property);
              break;
          }
        } else {
          component.addProperty(property);
        }

        this.currentLine++;
      }
    }

    throw new ParseException(`Unexpected end of input while parsing ${component.componentType}`);
  }

  private parseProperty(line: string): VCardProperty {
    const colonIndex = this.findUnquotedChar(line, ':');
    if (colonIndex === -1) {
      throw new ParseException(`Invalid property line (missing colon): ${line}`);
    }

    const nameAndParams = line.substring(0, colonIndex);
    let value = line.substring(colonIndex + 1);
    value = this.unescapeValue(value);

    const semicolonIndex = this.findUnquotedChar(nameAndParams, ';');
    let propertyName: string;
    let paramsPart: string | null = null;

    if (semicolonIndex !== -1) {
      propertyName = nameAndParams.substring(0, semicolonIndex).toUpperCase();
      paramsPart = nameAndParams.substring(semicolonIndex + 1);
    } else {
      propertyName = nameAndParams.toUpperCase();
    }

    const property = new VCardProperty(propertyName, value);

    // Validate property name (must be alphanumeric, hyphens, dots for groups)
    // Group names cannot have consecutive dots
    if (!/^[A-Za-z0-9._-]+$/.test(propertyName) || /\.\./.test(propertyName)) {
      throw new ParseException(`Invalid property name: ${propertyName}`);
    }

    if (paramsPart !== null) {
      this.parseParameters(paramsPart, property);
    }

    return property;
  }

  private parseParameters(paramsPart: string, property: VCardProperty): void {
    const parameters = this.splitParameters(paramsPart);

    for (const param of parameters) {
      const equalsIndex = param.indexOf('=');
      if (equalsIndex === -1) {
        throw new ParseException(`Invalid parameter (missing equals): ${param}`);
      }

      const paramName = param.substring(0, equalsIndex).toUpperCase();
      let paramValue = param.substring(equalsIndex + 1);

      // Remove quotes if present; don't split quoted values on commas
      let wasQuoted = false;
      if (paramValue.startsWith('"') && paramValue.endsWith('"') && paramValue.length >= 2) {
        paramValue = paramValue.substring(1, paramValue.length - 1);
        wasQuoted = true;
      }

      if (wasQuoted) {
        property.addParameter(paramName, paramValue);
      } else {
        const values = this.splitParameterValues(paramValue);
        for (const value of values) {
          property.addParameter(paramName, value);
        }
      }
    }
  }

  private splitParameters(paramsPart: string): string[] {
    const parameters: string[] = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < paramsPart.length; i++) {
      const c = paramsPart[i];
      if (c === '"') {
        inQuotes = !inQuotes;
        current += c;
      } else if (c === ';' && !inQuotes) {
        parameters.push(current);
        current = '';
      } else {
        current += c;
      }
    }

    if (current.length > 0) {
      parameters.push(current);
    }

    return parameters;
  }

  private splitParameterValues(paramValue: string): string[] {
    const values: string[] = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < paramValue.length; i++) {
      const c = paramValue[i];
      if (c === '"') {
        inQuotes = !inQuotes;
      } else if (c === ',' && !inQuotes) {
        values.push(current);
        current = '';
      } else {
        current += c;
      }
    }

    if (current.length > 0) {
      values.push(current);
    }

    return values;
  }

  private findUnquotedChar(str: string, target: string): number {
    let inQuotes = false;
    for (let i = 0; i < str.length; i++) {
      if (str[i] === '"') {
        inQuotes = !inQuotes;
      } else if (str[i] === target && !inQuotes) {
        return i;
      }
    }
    return -1;
  }

  private unescapeValue(value: string): string {
    return value
      .replace(/\\n/gi, '\n')
      .replace(/\\;/g, ';')
      .replace(/\\,/g, ',')
      .replace(/\\\\/g, '\\');
  }

  private validateVCard(vcard: VCardObject): void {
    if (!vcard.version) {
      throw new ParseException(
        'Missing required VERSION property (RFC 6350 Section 6.7.9). vCard must include VERSION:4.0',
      );
    }

    if (vcard.version !== '4.0') {
      throw new ParseException(
        `Unsupported vCard version: ${vcard.version}. Only version 4.0 is supported (see ADR 0004).`,
      );
    }

    if (!vcard.formattedName) {
      throw new ParseException(
        'Missing required FN (Formatted Name) property (RFC 6350 Section 6.2.1). vCard must include FN property.',
      );
    }
  }
}
